bin/improver-nowcast-optical-flow

```diff
@@ -0,0 +1,141 @@
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to calculate optical flow advection velocities with option to
+extrapolate."""
+
+import datetime
+import os
+import iris
+import numpy as np
+
+from improver.argparser import ArgParser
+from improver.nowcasting.optical_flow import AdvectField, OpticalFlow
+from improver.utilities.filename import generate_file_name
+from improver.utilities.load import load_cube
+from improver.utilities.save import save_netcdf
+
+
+def main():
+    """Calculate optical flow advection velocities and (optionally)
+    extrapolate data."""
+
+    parser = ArgParser(
+        description="Calculate optical flow components from input fields "
+        "and (optionally) extrapolate to required lead times.")
+
+    parser.add_argument("input_filepaths", metavar="INPUT_FILEPATHS",
+                        nargs="+", type=str, help="Paths to the input radar "
+                        "files.  There should be 3 input files at T, T-1 and "
+                        "T-2 from which to calculate optical flow velocities.")
+    parser.add_argument("--output_dir", metavar="OUTPUT_DIR", type=str,
+                        default='', help="Directory to write output files.")
+
+    # OpticalFlow plugin configurable parameters
+    parser.add_argument("--ofc_box_size", type=int, default=30, help="Size of "
+                        "square 'box' (in grid squares) within which to solve "
+                        "the optical flow equations.")
+    parser.add_argument("--smart_smoothing_iterations", type=int, default=100,
+                        help="Number of iterations to perform in enforcing "
+                        "smoothness constraint for optical flow velocities.")
+
+    # AdvectField options
+    parser.add_argument("--extrapolate", action="store_true", default=False,
+                        help="Optional flag to advect current data forward to "
+                        "specified lead times.")
+    parser.add_argument("--max_lead_time", type=int, default=360,
+                        help="Maximum lead time required (mins).  Ignored "
+                        "unless '--extrapolate' is set.")
+    parser.add_argument("--lead_time_interval", type=int, default=15,
+                        help="Interval between required lead times (mins). "
+                        "Ignored unless '--extrapolate' is set.")
+
+    args = parser.parse_args()
+
+    # read input data
+    cube_list = []
+    for fname in args.input_filepaths:
+        cube_list.append(load_cube(fname))
+    if len(cube_list) != 3:
+        raise ValueError("Expected exactly 3 input files.")
+
+    # order input files by validity time
+    cube_list.sort(key=lambda x: x.coord("time").points[0])
+    time_coord = cube_list[-1].coord("time")
+
+    # calculate optical flow velocities from T-1 to T and T-2 to T-1
```

Comment:
The ticket says use 3 fields to create an average of advection velocity fields but any reason why this should be extended to any number between 2 and n?

Reply:
From a technical perspective, only 2 input files (1 derived advection velocity field from T-1 to T) are needed for the process.  It helps to use 3 input files (2 velocity fields going back to T-2) to make the outputs stable and to increase the area over which velocities are available.  However the more files are input, the further back in time (T-3, T-4, etc) the resulting advection velocities were applicable.  So there is a balance between having more velocity fields and how representative they are of the current conditions.